backends/p4tools/testgen/targets/bmv2/p4_asserts_parser.cpp

```diff
@@ -0,0 +1,623 @@
+#include "backends/p4tools/testgen/targets/bmv2/p4_asserts_parser.h"
+
+#include <boost/multiprecision/cpp_int.hpp>
+#include <boost/optional/optional_io.hpp>
+
+#include "backends/p4tools/common/core/z3_solver.h"
+#include "backends/p4tools/common/lib/ir.h"
+#include "lib/error.h"
+
+namespace P4Tools {
+
+namespace AssertsParser {
+
+AssertsParser::AssertsParser(std::vector<std::vector<const IR::Expression*>>& output)
+    : restrictionsVec(output) {
+    setName("Restrictions");
+}
+
+bool is_space(char c) noexcept {
+    switch (c) {
+        case ' ':
+        case '\t':
+        case '\r':
+            return true;
+        default:
+            return false;
+    }
+}
+
+std::ostream& operator<<(std::ostream& os, const Token::Kind& kind) {
+    static const char* const names[]{
+        "Text",         "True",       "False",          "LineStatementClose",
+        "Id",           "Number",     "Minus",          "Plus",
+        "Dot",          "FieldAcces", "MetadataAccess", "LeftParen",
+        "RightParen",   "Equal",      "NotEqual",       "GreaterThan",
+        "GreaterEqual", "LessThan",   "LessEqual",      "LNot",
+        "Colon",        "Semicolon",  "Conjunction",    "Disjunction",
+        "Implication",  "Slash",      "Percent",        "Shr",
+        "Shl",          "Mul",        "Comment",        "Unknown",
+        "EndString",    "End",
+    };
+    return os << names[static_cast<int>(kind)];
+}
+
+const IR::Expression* makeExpr(std::vector<const IR::Expression*> input) {
+    const IR::Expression* expr = nullptr;
```

Comment:
Can you simplify this function and add comments? It is hard to understand and comprehend. Rule of thumb: Segment this code into independent logical parts and move these parts into a helper function.